papis/commands/update.py

```diff
@@ -39,25 +120,223 @@
     :prog: papis update
 """
 
-from typing import List, Dict, Tuple, Optional, Any
+import ast
+from typing import Any, Dict, List, Optional, Tuple
 
 import click
 
-import papis.utils
-import papis.strings
+import papis.cli
+import papis.commands.doctor
+import papis.config
 import papis.document
 import papis.format
-import papis.cli
-import papis.importer
 import papis.git
+import papis.importer
 import papis.logging
+import papis.strings
+import papis.utils
 
 logger = papis.logging.get_logger(__name__)
 
-
-def run(document: papis.document.Document,
-        data: Optional[Dict[str, Any]] = None,
-        git: bool = False) -> None:
+KEY_TYPES = papis.commands.doctor.get_key_type_check_keys()
+
+
+def try_parsing_str(key: str, value: str) -> str:
+    """
+    Tries to parse the input string as a python expression.
+
+    :returns: The parsed input string if string is a python expression,
+        otherwise an unchanged string.
+    """
+    try:
+        value = ast.literal_eval(value)
+    except (SyntaxError, ValueError):
+        logger.debug("Value '%s' of key '%s' is not a python expression.", value, key)
+    return value
+
+
+def run_set(
+    document: papis.document.DocumentLike, to_set: List[Tuple[str, str]]
+) -> None:
+    """
+    Processes a list of `to_set` tuples and applies the resulting changes to the
+    input document. Each tuple is (KEY, VALUE) and results in setting the KEY to
+    the VALUE in the document.
+
+    :returns: A boolean indicating whether the update was successful.
+    """
+    for key, value in to_set:
+        value = papis.format.format(value, document, default=value)
+        value = try_parsing_str(key, value)
+        # NOTE: the previous version of `papis update` handled these special
+        #       cases. should we continue doing so? i've done so in the most
+        #       'stupid' way, allowing users to mess things up if they set
+        #       these fields to something other than strings. the reason is
+        #       mainly that otherwise it would almost make more sense to just
+        #       always check if the type is right. but doing that required code
+        #       that is really just there to allow the user to should themselves
+        #       in the foot. not sure what's best here.
+        if not value:
+            document.pop(key, None)
+            break
+        if key == "notes" and isinstance(value, str):
+            document[key] = papis.utils.clean_document_name(value)
+        elif key == "files" and isinstance(value, list):
+            for file in value:
+                if isinstance(file, str):
+                    document[key].append(papis.utils.clean_document_name(file))
+                else:
+                    document[key].append(value)
```

Comment:
I thought there might be the case where you've renamed a file and want to adjust the info.yaml to reflect this. Although admittedly, the `clean_document_name` part would then probably not be all that useful (I still think it might be good to keep just be consistent re file names).

Reply:
Yeah, you'll definitely want to rename the file on disk as well here!